File: server/room-rules.js
// LUẬT XẾP PHÒNG — MỘT NƠI DUY NHẤT.
// Trước đây luật chỉ nằm trên giao diện: server ai gửi gì cũng nhận, và mỗi đường
// (tạo mới / sửa hồ sơ / chuyển phòng / check-in lại / duyệt đơn) lại hành xử một kiểu.
// Mọi đường xếp phòng PHẢI gọi checkRoomAssignment() ở đây.
const { query } = require('./db');

// Lỗi CHẶN (errors)  -> trả 400, không cho lưu.
// CẢNH BÁO (warnings) -> cho phép lưu NHƯNG người xếp phải xác nhận, và được GHI VẾT.
//
// Quá tải KHÔNG chặn: nghiệp vụ cho phép HV mới vào ở chờ HV cũ xuất cảnh (chốt 15/07/2026).
async function checkRoomAssignment({ studentId, gender, rentalType, roomId }) {
  const errors = [], warnings = [];
  if (!roomId) return { errors, warnings, room: null };

  const room = (await query(
    `SELECT id, name, gender, capacity, COALESCE(room_type,'shared') AS room_type FROM rooms WHERE id=$1 AND deleted_at IS NULL`,
    [roomId])).rows[0];
  if (!room) { errors.push('Phòng không tồn tại hoặc đã bị xoá'); return { errors, warnings, room: null }; }

  // 1) Giới tính — KHÔNG có ngoại lệ hợp lệ nào cho việc nam ở phòng nữ
  if (gender && room.gender && gender !== room.gender) {
    errors.push(`Phòng ${room.name} là phòng ${room.gender === 'female' ? 'NỮ' : 'NAM'}, không xếp được học viên ${gender === 'female' ? 'nữ' : 'nam'}`);
  }

  // Những người đang ở phòng (không tính chính học viên này khi sửa hồ sơ)
  const others = (await query(
    `SELECT id, name, rental_type FROM students
     WHERE room_id=$1 AND deleted_at IS NULL AND status='in' AND ($2::int IS NULL OR id <> $2)`,
    [roomId, studentId || null])).rows;

  // 2) Thuê nguyên phòng — 1 phòng chỉ 1 người thuê nguyên, và không ở ghép chung
  const nguyenPhong = others.find(o => o.rental_type === 'phong');
  if (rentalType === 'phong' && others.length) {
    errors.push(`Phòng ${room.name} đang có ${others.length} người ở — không thể cho thuê NGUYÊN PHÒNG. (Thuê nguyên phòng = thu trọn giá phòng; để 2 người cùng thuê nguyên phòng là thu 2 lần tiền cho 1 phòng.)`);
  } else if (rentalType !== 'phong' && nguyenPhong) {
    errors.push(`Phòng ${room.name} đã cho ${nguyenPhong.name} thuê NGUYÊN PHÒNG — không xếp thêm người ở ghép.`);
  }

  // 3) Quá tải — CẢNH BÁO, không chặn
  if (room.room_type === 'shared' && room.capacity > 0) {
    const sau = others.length + 1;
    if (sau > room.capacity) {
      warnings.push({
        code: 'OVER_CAPACITY',
        message: `Phòng ${room.name} đã đủ ${others.length}/${room.capacity} chỗ — xếp thêm sẽ thành QUÁ TẢI ${sau}/${room.capacity} (vượt ${sau - room.capacity} người).`,
        room_name: room.name, occupancy_after: sau, capacity: room.capacity, over_by: sau - room.capacity,
      });
    }
  }
  return { errors, warnings, room };
}

// Ghi vết riêng cho việc xếp gây QUÁ TẢI: ai xếp · lúc nào · HV nào · phòng nào · vượt mấy người.
// Ghi vào nhật ký hệ thống với dấu [QUÁ TẢI] để tra cứu được, không lẫn vào nhật ký chung.
async function logOverload(req, { studentId, studentName, warning }) {
  if (!warning || warning.code !== 'OVER_CAPACITY') return;
  const detail = `[QUÁ TẢI] Xếp học viên ${studentName || '#' + studentId} vào phòng ${warning.room_name} — ` +
    `${warning.occupancy_after}/${warning.capacity}, vượt ${warning.over_by} người`;
  try {
    await query(
      `INSERT INTO audit_log (user_id, username, role, method, path, detail) VALUES ($1,$2,$3,$4,$5,$6)`,
      [req.user?.id || null, req.user?.username || '', req.user?.role || '', req.method, (req.originalUrl || '').split('?')[0], detail]);
  } catch (e) { /* ghi vết hỏng không được chặn nghiệp vụ */ }
}

// Dùng trong route: trả về response 400/409 nếu cần, hoặc null nếu được phép đi tiếp.
// - Có lỗi chặn        -> 400
// - Có cảnh báo mà chưa xác nhận -> 409 kèm danh sách cảnh báo (client hiện hộp xác nhận rồi gửi lại confirm_overload:true)
function blockOrConfirm(res, { errors, warnings }, confirmed) {
  if (errors.length) { res.status(400).json({ error: errors.join(' · ') }); return true; }
  if (warnings.length && !confirmed) {
    res.status(409).json({
      error: warnings.map(w => w.message).join(' · '),
      needs_confirm: true, warnings,
      hint: 'Gửi lại kèm "confirm_overload": true để xác nhận vẫn xếp. Việc này sẽ được ghi vào nhật ký.',
    });
    return true;
  }
  return false;
}

module.exports = { checkRoomAssignment, logOverload, blockOrConfirm };

File: internal/roomrules/roomrules.go
// Package roomrules — LUẬT XẾP PHÒNG, một nơi duy nhất. Port từ server/room-rules.js.
// Lỗi CHẶN -> 400; CẢNH BÁO (quá tải) -> 409 cần xác nhận + ghi vết. Quá tải KHÔNG chặn (chốt 15/07/2026).
package roomrules

import (
	"context"
	"errors"
	"fmt"

	"github.com/jackc/pgx/v5"
	"ktx/internal/db"
)

type Warning struct {
	Code           string `json:"code"`
	Message        string `json:"message"`
	RoomName       string `json:"room_name"`
	OccupancyAfter int    `json:"occupancy_after"`
	Capacity       int    `json:"capacity"`
	OverBy         int    `json:"over_by"`
}

type RoomInfo struct {
	ID       int
	Name     string
	Gender   string
	Capacity int
	RoomType string
}

type Result struct {
	Errors   []string
	Warnings []Warning
	Room     *RoomInfo
}

// CheckRoomAssignment: kiểm luật xếp phòng. server/room-rules.js:11-51
func CheckRoomAssignment(ctx context.Context, q db.Querier, studentID *int, gender, rentalType string, roomID *int) (*Result, error) {
	res := &Result{Errors: []string{}, Warnings: []Warning{}}
	if roomID == nil {
		return res, nil
	}

	var room RoomInfo
	err := q.QueryRow(ctx,
		"SELECT id, name, gender, capacity, COALESCE(room_type,'shared') AS room_type FROM rooms WHERE id=$1 AND deleted_at IS NULL", *roomID).
		Scan(&room.ID, &room.Name, &room.Gender, &room.Capacity, &room.RoomType)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			res.Errors = append(res.Errors, "Phòng không tồn tại hoặc đã bị xoá")
			return res, nil
		}
		return nil, err
	}
	res.Room = &room

	// 1) Giới tính
	if gender != "" && room.Gender != "" && gender != room.Gender {
		g1 := "NAM"
		if room.Gender == "female" {
			g1 = "NỮ"
		}
		g2 := "nam"
		if gender == "female" {
			g2 = "nữ"
		}
		res.Errors = append(res.Errors, fmt.Sprintf("Phòng %s là phòng %s, không xếp được học viên %s", room.Name, g1, g2))
	}

	// Người đang ở (loại chính HV này khi sửa hồ sơ)
	rows, err := q.Query(ctx,
		`SELECT id, name, rental_type FROM students
		  WHERE room_id=$1 AND deleted_at IS NULL AND status='in' AND ($2::int IS NULL OR id <> $2)`,
		*roomID, studentID)
	if err != nil {
		return nil, err
	}
	type other struct {
		name   string
		rental string
	}
	var others []other
	for rows.Next() {
		var oid int
		var o other
		if err := rows.Scan(&oid, &o.name, &o.rental); err != nil {
			rows.Close()
			return nil, err
		}
		others = append(others, o)
	}
	rows.Close()
	if err := rows.Err(); err != nil {
		return nil, err
	}

	// 2) Thuê nguyên phòng
	var nguyenPhong *other
	for i := range others {
		if others[i].rental == "phong" {
			nguyenPhong = &others[i]
			break
		}
	}
	if rentalType == "phong" && len(others) > 0 {
		res.Errors = append(res.Errors, fmt.Sprintf("Phòng %s đang có %d người ở — không thể cho thuê NGUYÊN PHÒNG. (Thuê nguyên phòng = thu trọn giá phòng; để 2 người cùng thuê nguyên phòng là thu 2 lần tiền cho 1 phòng.)", room.Name, len(others)))
	} else if rentalType != "phong" && nguyenPhong != nil {
		res.Errors = append(res.Errors, fmt.Sprintf("Phòng %s đã cho %s thuê NGUYÊN PHÒNG — không xếp thêm người ở ghép.", room.Name, nguyenPhong.name))
	}

	// 3) Quá tải — CẢNH BÁO
	if room.RoomType == "shared" && room.Capacity > 0 {
		sau := len(others) + 1
		if sau > room.Capacity {
			res.Warnings = append(res.Warnings, Warning{
				Code:           "OVER_CAPACITY",
				Message:        fmt.Sprintf("Phòng %s đã đủ %d/%d chỗ — xếp thêm sẽ thành QUÁ TẢI %d/%d (vượt %d người).", room.Name, len(others), room.Capacity, sau, room.Capacity, sau-room.Capacity),
				RoomName:       room.Name,
				OccupancyAfter: sau,
				Capacity:       room.Capacity,
				OverBy:         sau - room.Capacity,
			})
		}
	}
	return res, nil
}

// LogOverload: ghi vết việc xếp gây QUÁ TẢI vào audit_log. server/room-rules.js:55-64
func LogOverload(ctx context.Context, q db.Querier, userID *int, username, role, method, path string, studentID int, studentName string, w Warning) {
	if w.Code != "OVER_CAPACITY" {
		return
	}
	who := studentName
	if who == "" {
		who = fmt.Sprintf("#%d", studentID)
	}
	detail := fmt.Sprintf("[QUÁ TẢI] Xếp học viên %s vào phòng %s — %d/%d, vượt %d người",
		who, w.RoomName, w.OccupancyAfter, w.Capacity, w.OverBy)
	_, _ = q.Exec(ctx,
		"INSERT INTO audit_log (user_id, username, role, method, path, detail) VALUES ($1,$2,$3,$4,$5,$6)",
		userID, username, role, method, path, detail)
}
